Meta/CMake/serenity_components.cmake

```
@@ -0,0 +1,90 @@
+#
+# Target properties and functions for optionally building libraries and applications
+#
+
+define_property(TARGET PROPERTY SERENITY_COMPONENT_NAME BRIEF_DOCS "SerenityOS component name" FULL_DOCS "-")
+define_property(TARGET PROPERTY SERENITY_COMPONENT_DESCRIPTION BRIEF_DOCS "SerenityOS component description" FULL_DOCS "-")
+define_property(TARGET PROPERTY SERENITY_COMPONENT_RECOMMENDED BRIEF_DOCS "SerenityOS component recommended (flag)" FULL_DOCS "-")
+define_property(TARGET PROPERTY SERENITY_COMPONENT_REQUIRED BRIEF_DOCS "SerenityOS component required (flag)" FULL_DOCS "-")
+define_property(TARGET PROPERTY SERENITY_COMPONENT_DEPENDS BRIEF_DOCS "SerenityOS component dependencies" FULL_DOCS "-")
+
+function(serenity_component name)
+    cmake_parse_arguments(PARSE_ARGV 1 SERENITY_COMPONENT "RECOMMENDED;REQUIRED" "DESCRIPTION" "TARGETS;DEPENDS")
+    string(TOUPPER "${name}" NAME_UPPER)
+    option("BUILD_${NAME_UPPER}" "Build ${name}" "${SERENITY_COMPONENT_RECOMMENDED}")
+    add_custom_target("Component${name}")
+    set_target_properties(
+        "Component${name}" PROPERTIES
+        SERENITY_COMPONENT_NAME        "${name}"
+        SERENITY_COMPONENT_DESCRIPTION "${SERENITY_COMPONENT_DESCRIPTION}"
+        SERENITY_COMPONENT_RECOMMENDED "${SERENITY_COMPONENT_RECOMMENDED}"
+        SERENITY_COMPONENT_REQUIRED    "${SERENITY_COMPONENT_REQUIRED}"
+        SERENITY_COMPONENT_DEPENDS     "${SERENITY_COMPONENT_DEPENDS}"
+    )
+    if(NOT "${SERENITY_COMPONENT_TARGETS}" STREQUAL "") # note: one component is /bin/false, which makes this tricky
+        foreach(target IN LISTS SERENITY_COMPONENT_TARGETS)
+            add_dependencies("Component${name}" "${target}")
+        endforeach()
+    endif()
+    if(BUILD_EVERYTHING OR "BUILD_${NAME_UPPER}" OR SERENITY_COMPONENT_REQUIRED)
```

I just noticed, but this should be dereferenced here. See [CMP0054](https://cmake.org/cmake/help/latest/policy/CMP0054.html)
```suggestion
    if(BUILD_EVERYTHING OR "${BUILD_${NAME_UPPER}}" OR SERENITY_COMPONENT_REQUIRED)
```

---

Same on line 69 (nice) as well. See [if()](https://cmake.org/cmake/help/latest/command/if.html#basic-expressions) for how these would be evaluated.

---

oh! good catch. Yeah we want to check these vs "ON/OFF" rather than vs existence.